### ISSUE TITLE
Give IL Ag Teachers Access to New Library

### ISSUE THREAD

**Issue by dpentony** [MyCaert/Issues-Enhancements#161]
We've recently completed the Ag Power library. We need to give all IL Ag subscribers access to the lesson plans, Eunits and CSAT questions in that library. The expiration date should be 12/1/2017. See screenshot below. I've already changed it for my IL ag account. Please let us know when this is completed.

![image](https://cloud.githubusercontent.com/assets/5790325/9941436/f03500ba-5d3a-11e5-9627-005a777aa3ec.png)

**Comment by bguerrettaz** [comment]
This request is now complete.  Can you please test a few accounts
(including your own) to assure everything looks correct?

Thanks,
Blake

**Comment by dpentony** [comment]
Thanks. Looks good.

**Comment by dpentony** [comment]
Now that the library is available, I went in an built a course outline under my IL profile. The outline is called Ag Power 2. After I opened the saved outline, every lesson I chose is showing up twice in the outline. Why are they showing twice? I checked the the admin and we've only entered the lesson names once. This is only happening after the lessons are saved to a course outline.

**Comment by bguerrettaz** [comment]
Ok, this is resolved.  The problem was there were five teachers that
already had access to this CD (yours, Mindy's, etc.)  which had duplicates
after I did a bulk add.  I manually removed the duplicates from those five
teachers.

Blake

**Comment by dpentony** [comment]
Looks good. Thanks.

**Comment by dpentony** [comment]
URGENT!!! Something happened as a result of giving all IL teachers access to the new AG Power library lessons, Eunits, and CSAT questions. It seems when they were given access to the new library, it took away access for many teachers to the CSAT questions for all other libraries. All it left was dates for the Ag Power library. It wiped out the CSAT activate/deactivate dates for all other libraries. Please advise. This issue is widespread. All activate/deactivate dates need to be replaced for the CSAT questions and be the same as the dates for the library's lessons & Eunits. See a sample below:

![image](https://cloud.githubusercontent.com/assets/5790325/10140488/5f4aa58c-65cd-11e5-81be-a2d032eb6fd0.png)

**Comment by bguerrettaz** [comment]
Ok, I found the issue, when the SQL statement to update the CSATArray list
was executed on the 21st (when this change was made initially), it was a
corrupted statement, meaning, it updated all of those teachers' column with
blanks.  I have no idea what caused this as the same exact script was used
for Lessons and eUnits and worked correctly.  For whatever reason, my guess
is there was an interrupt during the statement execution which failed the
update.

It looks like this wasn't caught in testing, but instead today.

To 'ease the pain' I created a script to take today's version (instead of a
7 day old backup) and updated the CSATArray with the correct entries.  The
option was to take the backup from the 21st, losing any change between
today and then.

Bottom line is I have executed the script and looked through the teachers,
all seem to be correct now.  Please test and confirm this with the teachers
you have tested, etc.

Again - this seems to be a corrupted (interrupted) statement execution.
The same script was used on the other entities and was used in part today
correctly.

Please let me know if you have any questions.

Thanks,
Blake

**Comment by dpentony** [comment]
Three out of five teachers I randomly checked are still without most CSAT access. Please take a look at the accounts for Terry Lancaster, Julie Niemerg, and Scott Zehr. All of them still only have access to CSAT for Ag Power.

**Comment by bguerrettaz** [comment]
I see it now, some of the records on the CSAT Activation date table has the
same issue.  For this table, I'll have to restore and reapply the new
Library.  There's no way to save changes from the 21st until today.

**Comment by dpentony** [comment]
Please let us know when you have done so and we will re-test.

**Comment by bguerrettaz** [comment]
I was incorrect about this table.  CSAT is actually different than
Lessons/eUnits.  When I restored the old table, everything is looking
correct.  When I dug a little more, about half of the teachers had no
records in that table.  The code actually defaults to the teacher record if
there are no records in the CSAT Activation Dates table, which is why all
of this works.

I'd like to get this corrected, but I feel that it would be quite a process
to get all three tables aligned correctly.

To correct this the way it is, for the half that DO have records in this
table, they need to have Ag Power library added, for those that don't, its
ready to go.

So I'll get the ones added that need to be for this to work and let you
know.

**Comment by bguerrettaz** [comment]
Ok, records are inserted, all should be showing to the user as normal.

For reference:

insert into tblCSATActivationDates (State, TeacherID, CDID, ActivateDate,
DeactivateDate)
select 'IL' as State, teacherID, 85 as CDID, '09/21/2015' as ActivateDate,
'12/01/2017' as DeactivateDate from
(
select distinct TeacherID from tblCSATActivationDates where TeacherID in
(
select TeacherID from tblTeacher t
join tblSchool s on t.SchoolID = s.SchoolID
where s.State = 'IL' and t.Status = 'Active' and t.Focus = 'Ag'
)
)a

**Comment by dpentony** [comment]
Blake: you said "The code actually defaults to the teacher record if there are no records in the CSAT Activation Dates table, which is why all of this works." 

How did those CSAT activation dates disappear in the first place?

As a heads up, we'll be adding another new library for all IL teachers in about a month. It will be the Ag, Food and Natural Resources library. All ag teachers will get access to the library's lessons, Eunits and CSAT questions. When it comes time to add it, let's remember the issues we had with adding the Ag Power.

**Comment by bguerrettaz** [comment]
According to the table history, CSAT wasn't ever correct like the Lessons
and eUnits were.  CSAT was always unique and always 'appeared' correct to
the user because of the defaulting back to the teacher record, instead of
the CSATActivationDate record.

I will remember this uniqueness going forward for future bulk
changes/additions.

Blake

**Comment by dpentony** [comment]
I am not sure this is totally solved. We had a couple of Illinois teachers report that they only had access to the lessons and Eunits for one library (Ag Power), but they had access to the CSAT questions for ALL libraries. See screenshot below. This is what Trent Taber's profile looked like before I fixed it. I had to give him back access to the lessons and Eunits for the libraries that he only had CSAT access for.

![image](https://cloud.githubusercontent.com/assets/5790325/10152206/6fb6d52e-6615-11e5-8f3f-7e686376af72.png)

**Comment by bguerrettaz** [comment]
Lesson Plan and eUnit database tables had a similar, but much smaller
problem. (89 teachers total)

I altered their records so their edit teacher profile should now be correct.

Going forward with bulk adds or changes, there will be some steps I'll have
to take to cater to both methods (its kind of an all or nothing in the
tables), so I'll just have to account for that.

Blake

**Comment by mnichols13** [comment]
found another teacher who had her libraries wiped out - Erica James at Eldorado

![image](https://cloud.githubusercontent.com/assets/5790342/10174482/a418f546-66af-11e5-8488-b1a76501f556.png)

**Comment by mnichols13** [comment]
also Kathy Novotney at FCAE school.  I left hers as is...

**Comment by bguerrettaz** [comment]
This should be fixed now.  More cleanup from the teacher record
perspective.  I did it for all teachers I found with the same issue.

**Comment by dpentony** [comment]
We've got another new library we've added to MYcaert. We need to give all IL ag teachers access to the Agriculture, Food, and Natural Resources library  until 12/1/2017. Teachers should be given access to the lesson plans, Eunits, and CSAT questions for the library. See screenshot below:

![image](https://cloud.githubusercontent.com/assets/5790325/11154703/a06c4464-8a06-11e5-8f89-e1521a8cf436.png)

@bguerrettaz - please note the various issues we encountered (above) when we did this same thing for the Ag Power library.I'd like to avoid those issues this time.

**Comment by bguerrettaz** [comment]
I wanted to let you know that I'm working on this as well, I want to make sure its done correctly using the past experience as leverage.  I should have it done soon.

Blake

**Comment by bguerrettaz** [comment]
Ok, this took a bit longer than the last because I wanted to try as much as possible to have little to no issues with this upgrade, taking into account the problems from last time.

I have backups of all 4 tables impacted in case of any issues.  I have tested the accounts from above in this issue as well as a few others and all looks good to me.

Please test and confirm this upgrade as much as possible so we feel good about this upgrade.

Below is the SQL i had to write for this, its a very manual process, so the below is mostly for my reference, although each one is unique.

For reference:

--Find new CD for this update
select * from tblCD where CDID = 87

--Backup tblTeacher
select * into tblTeacher_111715 from tblTeacher
--Get list of teachers that need their teacher record altered with 87
select t.teacherID, t.LessonPlanArray, t.eUnitArray, t.CSATArray from tblTeacher t 
join tblSchool s on t.SchoolID = s.SchoolID
where s.State = 'IL' and t.Focus = 'Ag' and t.Status = 'Active' and t.TeacherID NOT IN
(
14083,
14060,
13429,
11251,
3369,
3364,
16110
) 

--Update tblTeacher record for each teacher from above
update tblTeacher set 
LessonPlanArray = LessonPlanArray + ', 87',
eUnitArray = eUnitArray + ', 87',
CSATArray = CSATArray + ', 87'
where TeacherID in
(
select distinct t.teacherID from tblTeacher t 
join tblSchool s on t.SchoolID = s.SchoolID
where s.State = 'IL' and t.Focus = 'Ag' and t.Status = 'Active' and t.TeacherID NOT IN
(
14083,
14060,
13429,
11251,
3369,
3364,
16110
) 
)



--Backup tblLessonPlanActivationDates table
select * into tblLessonPlanActivationDates_111715 from tblLessonPlanActivationDates
--Insert into LPAD table
insert into tblLessonPlanActivationDates (State, TeacherID, CDID, ActivateDate,
DeactivateDate)
select 'IL' as State, a.teacherID, 87 as CDID, '11/17/2015' as ActivateDate, 
'12/01/2017' as DeactivateDate from
(
select TeacherID from tblLessonPlanActivationDates where TeacherID IN
(
	select distinct t.teacherID from tblTeacher t 
	join tblSchool s on t.SchoolID = s.SchoolID
	where s.State = 'IL' and t.Focus = 'Ag' and t.Status = 'Active' and t.TeacherID NOT IN
	(
	14083,
	14060,
	13429,
	11251,
	3369,
	3364,
	16110
	)
) and CDID = 85
)a


--Backup tblEunitActivationDates table
select * into tblEunitActivationDates_111715 from tblEunitActivationDates
--Insert into EUAD table
insert into tblEunitActivationDates (State, TeacherID, CDID, ActivateDate,
DeactivateDate)
select 'IL' as State, a.teacherID, 87 as CDID, '11/17/2015' as ActivateDate, 
'12/01/2017' as DeactivateDate from
(
select TeacherID from tblEunitActivationDates where TeacherID IN
(
	select distinct t.teacherID from tblTeacher t 
	join tblSchool s on t.SchoolID = s.SchoolID
	where s.State = 'IL' and t.Focus = 'Ag' and t.Status = 'Active' and t.TeacherID NOT IN
	(
	14083,
	14060,
	13429,
	11251,
	3369,
	3364,
	16110
	)
) and CDID = 85
)a



--Backup tblCSATActivationDates table
select * into tblCSATActivationDates_111715 from tblCSATActivationDates
--Insert into CSAD table
insert into tblCSATActivationDates (State, TeacherID, CDID, ActivateDate,
DeactivateDate)
select 'IL' as State, a.teacherID, 87 as CDID, '11/17/2015' as ActivateDate, 
'12/01/2017' as DeactivateDate from
(
select TeacherID from tblCSATActivationDates where TeacherID IN
(
	select distinct t.teacherID from tblTeacher t 
	join tblSchool s on t.SchoolID = s.SchoolID
	where s.State = 'IL' and t.Focus = 'Ag' and t.Status = 'Active' and t.TeacherID NOT IN
	(
	14083,
	14060,
	13429,
	11251,
	3369,
	3364,
	16110
	)
) and CDID = 85
)a


--The following teachers already have CD 87 or were blank for LessonPlanArray
--14083
--14060
--13429
--11251
--3369
--3364
--16110

**Comment by mnichols13** [comment]
looks good as far as i can see.  I looked at several IL teachers

**Comment by mnichols13** [comment]
Teachers are unable to access E-units in A1, they get the screen that pops up when a subscription doesn't include E-units that allows them to view samples instead.
![image](https://cloud.githubusercontent.com/assets/5790342/11253457/45ebdd70-8e01-11e5-9f3b-3ad6280408dc.png)

**Comment by mnichols13** [comment]
so far every teacher I have checked - about 10.  Interestingly enough even on my account - which would not have been updated with the rest of the teachers.  I gave myself access a few months back.  So I am not really sure that this is a problem related to this particular thread.

**Comment by bguerrettaz** [comment]
That's a good call out, it might not be related, I'll dig into this issue.

**Comment by bguerrettaz** [comment]
I did some research and found the root cause of the issue.  Please see the screenshot below.  This shows the database extract for the new library:

![image](https://cloud.githubusercontent.com/assets/5695475/11260595/55ea5c48-8e32-11e5-91ce-c87b78698294.png)

Somehow, the units that were entered were entered twice, once for 86 and once for 87.  The 87 is correct as its the new library CDID (CD identifier).

How are the new units entered?  I can't remember exactly.  The fix would be to remove all of the records that have an 86 in the CDID column in the above screenshot.  Once you do that, everything will work correctly.

Please let me know if you have questions and confirm that you agree to remove the duplicates (the 86's).

Thanks,
Blake

**Comment by dpentony** [comment]
Where is the screen shot and data table coming from? It's not familiar to me. Is that from the .csv file when we upload the actual lesson files? I am trying to understand where we would go to delete the 86 records.

**Comment by bguerrettaz** [comment]
This screenshot is from the database.  I would have to delete the records.
I'm guessing it came from the .csv upload process.

Thanks,
Blake

**Comment by dpentony** [comment]
I will talk to our people back at the office to determine what's going on.

**Comment by dpentony** [comment]
We will probably have you delete the extra records, but  I just want to talk to the office to try to determine why they seem to have been uploaded twice.

**Comment by dpentony** [comment]
@bguerrettaz  - I guess we will need you to delete the 86 series files. We still can't determine how duplicate  files exist. When you look at the Missing File report for the  AFNR library, no duplicates show up:

![image](https://cloud.githubusercontent.com/assets/5790325/11273488/2a928b4e-8e99-11e5-9402-0cdc30ef9aff.png)

If you can delete the duplicate files, great. If we need to talk in order to get this straightened out, please call my cell at 217.597.9303.

Just a thought. Are we  confusing the AFNR library with the IN Ag Food and Natural Resource library  (IANF)?

**Comment by mnichols13** [comment]
Dan,
see here on this screenshot
![image](https://cloud.githubusercontent.com/assets/5790342/11273664/26ab1fcc-8e9a-11e5-88ec-0e91a7e6d1eb.png)

**Comment by bguerrettaz** [comment]
I can just delete the records, I'll do that and let you know.

**Comment by bguerrettaz** [comment]
I have deleted these records and tested.  This issue is now resolved.

Please test and confirm.

**Comment by mnichols13** [comment]
looks good.
